src/main/java/com/microsoft/graph/serializer/DefaultSerializer.java

```diff
@@ -118,59 +118,70 @@ public DefaultSerializer(final ILogger logger) {
+    private void setChildAdditionalData(final IJsonBackedObject serializedObject, final JsonObject rawJson) {
         // Use reflection to iterate through fields for eligible Graph children
-        for (java.lang.reflect.Field field : serializedObject.getClass().getFields()) {
-            try {
-                Object fieldObject = field.get(serializedObject);
-
-                // If the object is a HashMap, iterate through its children
-                if (fieldObject instanceof HashMap) {
-                    HashMap<String, Object> serializableChildren = (HashMap<String, Object>) fieldObject;
-                    Iterator<Entry<String, Object>> it = serializableChildren.entrySet().iterator();
+        if(rawJson != null) {
+            for (java.lang.reflect.Field field : serializedObject.getClass().getFields()) {
+                try {
+                    if(field != null) {
+                        final Object fieldObject = field.get(serializedObject);
+                        if (fieldObject instanceof HashMap) {
+                            // If the object is a HashMap, iterate through its children
+                            @SuppressWarnings("unchecked")
+                            final HashMap<String, Object> serializableChildren = (HashMap<String, Object>) fieldObject;
+                            final Iterator<Entry<String, Object>> it = serializableChildren.entrySet().iterator();
 
-                    while (it.hasNext()) {
-                        Map.Entry<String, Object> pair = (Map.Entry<String, Object>)it.next();
-                        Object child = pair.getValue();
+                            while (it.hasNext()) {
+                                final Map.Entry<String, Object> pair = (Map.Entry<String, Object>)it.next();
+                                final Object child = pair.getValue();
 
-                        // If the item is a valid Graph object, set its additional data
-                        if (child instanceof IJsonBackedObject) {
-                            AdditionalDataManager childAdditionalDataManager = ((IJsonBackedObject) child).additionalDataManager();
-                            if(rawJson != null && field != null && rawJson.get(field.getName()) != null && rawJson.get(field.getName()).isJsonObject() 
-                            		&& rawJson.get(field.getName()).getAsJsonObject().get(pair.getKey()).isJsonObject()) {
-                            	childAdditionalDataManager.setAdditionalData(rawJson.get(field.getName()).getAsJsonObject().get(pair.getKey()).getAsJsonObject());
-                            	setChildAdditionalData((IJsonBackedObject) child,rawJson.get(field.getName()).getAsJsonObject().get(pair.getKey()).getAsJsonObject());
+                                // If the item is a valid Graph object, set its additional data
+                                if (child instanceof IJsonBackedObject) {
```

Comment:
I'd like to see a test in case where the child is a OData primitive, do we add non-schematized fields to additionalData?

Reply:
tweaked the last unit test I added so it tests for that as well.